### PR TITLE
Add GraniteMoeHybrid support for 4.0

### PR DESCRIPTION
**Pull request by Ssukriti**
# What does this PR do?

The PR adds support for upcoming Granite4.0 models. It terms of model architecture, it is a hybrid class with shared MLP layer and Bamba layers.

Fixes # (issue)

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

<!-- Your PR will be replied to more quickly if you can figure out the right person to tag with @

 If you know how to use git blame, that is the easiest way, otherwise, here is a rough guide of **who to tag**.
 Please tag fewer than 3 people.

Models:

### CODE REVIEW

**Comment by Rocketknight1** [comment]
cc @arthurzucker for text models!

File: src/transformers/models/granitemoehybrid/modular_granitemoehybrid.py [hunk 1]
```diff
@@ -0,0 +1,505 @@
...
+        super().__init__(config, layer_idx)
+
+class GraniteMoeHybridFlashAttention2(GraniteMoeSharedFlashAttention2):
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+
+class GraniteMoeHybridSdpaAttention(GraniteMoeSharedSdpaAttention):
+    pass
+
+GRANITEMOEHYBRID_ATTENTION_CLASSES = {
```
Comment (vasqu) [inline src/transformers/models/granitemoehybrid/modular_granitemoehybrid.py:57]: Just as a heads up, I think it would be nice to follow using the new attention interface (see #35235 for the original PR). Llama can also provide a good first pointer for this, e.g. https://github.com/huggingface/transformers/blob/de182ba2690fe6c3466f6463c7f4b3a61694b885/src/transformers/models/llama/modeling_llama.py#L216

(Except I'm missing that this is a more special kind of attention here :D )

Comment (alex-jw-brooks) [reply src/transformers/models/granitemoehybrid/modular_granitemoehybrid.py:56]: Thanks for the heads up @vasqu! We are still cleaning up this branch a bit, will take a look at this once the tests are in a better state 🙂

File: src/transformers/models/granitemoe/modeling_granitemoe.py
Comment (ArthurZucker) [inline src/transformers/models/granitemoe/modeling_granitemoe.py:1]: nice updates

File: src/transformers/models/granitemoehybrid/modular_granitemoehybrid.py [hunk 2]
```diff
@@ -0,0 +1,528 @@
...
+                attention_mask=attention_mask,
+                past_key_value=past_key_value,
+                output_attentions=output_attentions,
+                use_cache=use_cache,
+                cache_position=cache_position,
+                position_embeddings=position_embeddings,
+                **kwargs,
+            )
+        else:
+            raise ValueError(f"Expected layer type in ['attention', 'mamba'], got {self.layer_type}")
```
Comment (ArthurZucker) [inline src/transformers/models/granitemoehybrid/modular_granitemoehybrid.py:139]: config check should be sufficient!

File: src/transformers/models/granitemoehybrid/modular_granitemoehybrid.py
Comment (ArthurZucker) [inline src/transformers/models/granitemoehybrid/modular_granitemoehybrid.py:1]: Very nice! Thanks for highlighting the key differences!

**Review by ArthurZucker — COMMENTED** [review]
Its already in a great state! Thanks a lot both 🤗

**Comment by ArthurZucker** [comment]
Feel free to ping me again!

File: src/transformers/models/granitemoehybrid/modular_granitemoehybrid.py (hunk 2 above)
Comment (alex-jw-brooks) [reply src/transformers/models/granitemoehybrid/modular_granitemoehybrid.py:137]: Good point! Moved it 😄

File: src/transformers/models/granitemoehybrid/modular_granitemoehybrid.py (hunk 1 above)
Comment (alex-jw-brooks) [reply src/transformers/models/granitemoehybrid/modular_granitemoehybrid.py]: Thanks for the pointer @vasqu! Refactored this PR to the new attention interface 😄

**Comment by alex-jw-brooks** [comment]
Thanks @ArthurZucker! It's ready for another look when you get the chance!

File: src/transformers/models/granitemoehybrid/modular_granitemoehybrid.py [hunk 3]
```diff
@@ -0,0 +1,526 @@
...
+                with `head_dim` being the embedding dimension of each attention head.
+            kwargs (`dict`, *optional*):
+                Arbitrary kwargs to be ignored, used for FSDP and other methods that injects code
+                into the model
+        """
+        residual = hidden_states
+
+        hidden_states = self.input_layernorm(hidden_states)
+        self_attn_weights = None
+        if self.layer_type == "mamba":
```
Comment (ArthurZucker) [inline src/transformers/models/granitemoehybrid/modular_granitemoehybrid.py:118]: I am thinking let's remove the check on type, rely rather on the check of self.self_attn is not None?

File: src/transformers/models/granitemoehybrid/modular_granitemoehybrid.py (hunk 2 above)
Comment (ArthurZucker) [reply src/transformers/models/granitemoehybrid/modular_granitemoehybrid.py:137]: still todo 😉

File: src/transformers/models/granitemoehybrid/modular_granitemoehybrid.py [hunk 4]
```diff
@@ -0,0 +1,526 @@
...
+            raise ValueError(f"Expected layer type in ['attention', 'mamba'], got {self.layer_type}")
+
+        hidden_states = residual + hidden_states * self.residual_multiplier
+
+        # Fully Connected
+        residual = hidden_states
+        hidden_states = self.post_attention_layernorm(hidden_states)
+        moe_hidden_states, router_logits = self.block_sparse_moe(hidden_states)
+
+        if self.shared_mlp is None:
```
Comment (ArthurZucker) [inline src/transformers/models/granitemoehybrid/modular_granitemoehybrid.py:146]: I don't know if you answered or not, is there two different checkpoint being released, one with / one without this?

File: src/transformers/models/granitemoehybrid/modular_granitemoehybrid.py [hunk 5]
```diff
@@ -0,0 +1,526 @@
...
+                layer_outputs = decoder_layer(
+                    hidden_states,
+                    attention_mask=layer_mask,
+                    past_key_value=past_key_values,
+                    output_attentions=output_attentions,
+                    use_cache=use_cache,
+                    cache_position=cache_position,
+                    output_router_logits=output_router_logits,
+                    position_embeddings=position_embeddings,
+                )
```
Comment (ArthurZucker) [inline src/transformers/models/granitemoehybrid/modular_granitemoehybrid.py:403]: let's use the new GradientCHeckpointingLayer wdyt?

File: src/transformers/models/granitemoehybrid/modular_granitemoehybrid.py [hunk 6]
```diff
@@ -0,0 +1,526 @@
...
+        hidden_states = self.norm(hidden_states)
+
+        # add hidden states from the last decoder layer
+        if output_hidden_states:
+            all_hidden_states += (hidden_states,)
+
+        next_cache = next_decoder_cache if use_cache else None
+
+        if not return_dict:
+            return tuple(v for v in [hidden_states, next_cache, all_hidden_states, all_self_attns] if v is not None)
```
Comment (ArthurZucker) [inline src/transformers/models/granitemoehybrid/modular_granitemoehybrid.py:429]: we have a `@can_return_tuple` for the forward
```suggestion
```

File: tests/models/granitemoehybrid/test_modeling_granitemoehybrid.py [hunk 7]
```diff
@@ -0,0 +1,528 @@
...
+    from transformers import (
+        GraniteMoeHybridForCausalLM,
+        GraniteMoeHybridModel,
+    )
+    from transformers.models.bamba.modeling_bamba import (
+        HybridMambaAttentionDynamicCache,
+    )
+
+
+class GraniteMoeHybridModelTester:
```
Comment (ArthurZucker) [inline tests/models/granitemoehybrid/test_modeling_granitemoehybrid.py:50]: can we try to inherit tests from closes model so mambda in the same fashion as here https://github.com/huggingface/transformers/blob/86d38f163fb453bf8e0e878316c7217143a6d1e2/tests/models/gemma2/test_modeling_gemma2.py#L51

**Review by ArthurZucker — APPROVED** [review]
Very nice use of modular thanks a lot! 🤗

**Review by berserkr — COMMENTED** [review]
https://github.com/huggingface/transformers/blob/8274d2c5e5ff4d1cebb9f60c72e3b0d417726b76/src/transformers/models/granitemoehybrid/modeling_granitemoehybrid.py#L1199 - `std` initialized twice - std = self.config.initializer_range

File: src/transformers/models/granitemoehybrid/modular_granitemoehybrid.py (hunk 6 above)
Comment (alex-jw-brooks) [reply src/transformers/models/granitemoehybrid/modular_granitemoehybrid.py]: Nice! Added

File: src/transformers/models/granitemoehybrid/modular_granitemoehybrid.py (hunk 5 above)
Comment (alex-jw-brooks) [reply src/transformers/models/granitemoehybrid/modular_granitemoehybrid.py]: Definitely, that is a lot cleaner! I updated the models in the chain for modular to all use the gradient checkpointing layer (GraniteMoe/GraniteMoeShared/GraniteMoeHybrid)

File: src/transformers/models/granitemoehybrid/modular_granitemoehybrid.py (hunk 4 above)
Comment (alex-jw-brooks) [reply src/transformers/models/granitemoehybrid/modular_granitemoehybrid.py:151]: The models that are about to come out do use it! I think there are likely experiments ongoing without it, but am not sure about concrete plans for when they'll be released since I'm not the one training the models 🙂

File: src/transformers/models/granitemoehybrid/modular_granitemoehybrid.py (hunk 3 above)
Comment (alex-jw-brooks) [reply src/transformers/models/granitemoehybrid/modular_granitemoehybrid.py]: I agree, I also didn't like `self.mamba` being conditionally undefined. Updated this to define both in `__init__` and just check do mama if `self.mamba is not None`  and attention otherwise 🙂

File: src/transformers/models/granitemoehybrid/modular_granitemoehybrid.py (hunk 2 above)
Comment (alex-jw-brooks) [reply src/transformers/models/granitemoehybrid/modular_granitemoehybrid.py]: Whoops, sorry! Fixed

**Comment by alex-jw-brooks** [comment]
Thanks @berserkr! There were two because of `modular` expanding the superclass implementation that also set it. Updated to just pass the config value directly so it's less weird looking 🙂

File: tests/models/granitemoehybrid/test_modeling_granitemoehybrid.py (hunk 7 above)
Comment (alex-jw-brooks) [reply tests/models/granitemoehybrid/test_modeling_granitemoehybrid.py]: Good idea! The closest models are for the tests are `bamba`. Consolidated a bit to use Bamba tests, should be way easier to look at now 🤞

**Comment by alex-jw-brooks** [comment]
Thank you very much for the fast review @ArthurZucker! I've made all the changes 🙂

File: tests/models/granitemoehybrid/test_modeling_granitemoehybrid.py (hunk 7 above)
Comment (ArthurZucker) [reply tests/models/granitemoehybrid/test_modeling_granitemoehybrid.py]: perfect!

File: src/transformers/models/granitemoehybrid/modular_granitemoehybrid.py (hunk 4 above)
Comment (ArthurZucker) [reply src/transformers/models/granitemoehybrid/modular_granitemoehybrid.py:151]: In that case lets remove what's uncertain! 🤗

**Review by ArthurZucker — APPROVED** [review]
Marvelous ! Merging once the build PR passes! (should be easy to fix!)

File: src/transformers/models/granitemoehybrid/modular_granitemoehybrid.py (hunk 4 above)
Comment (alex-jw-brooks) [reply src/transformers/models/granitemoehybrid/modular_granitemoehybrid.py]: Sounds good! Removed the case with 0 experts, I'll open a follow-up PR if it ends up being used in a model to be released 😄

**Comment by alex-jw-brooks** [comment]
Thanks @ArthurZucker! Added the missing TOC entry and removed the currently unused shared condition for the MLP, should pass now! 🤞
